README.md

```diff
@@ -130,29 +128,80 @@ Examples:
 
 ### Scopes
 
-Scopes should be specified in disjunctive normal form.  In other words, an
-array of arrays, where all of the scopes in at least one of the inner arrays
-must be satisfied.  Parameters are substituted into scopes with `<paramName>`
+Scopes should be specified in
+[scope expression form](https://github.com/taskcluster/taskcluster-lib-scopes#new-style).
+Parameters are substituted into scopes with `<paramName>`
 syntax.  For example, the following defintion allows the method when *either*
 the caller's scopes satisfy `queue:create-task..` for the given `provisionerId`
 and `workerType`, *or* the caller's scopes satisfy all of
 `queue:define-task:..`, `queue:task-group-id:..`, and `queue:schedule-task:..`.
 
 ```js
-  scopes: [
-    [
+  scopes: {AnyOf: [
+    {AllOf: [
       'queue:create-task:<provisionerId>/<workerType>',
-    ], [
+    ]},
+    {AllOf: [
       'queue:define-task:<provisionerId>/<workerType>',
       'queue:task-group-id:<schedulerId>/<taskGroupId>',
       'queue:schedule-task:<schedulerId>/<taskGroupId>/<taskId>',
-    ],
-  ],
+    ]},
+  ]},
 ```
 
 If scope validation fails, the user is presented with an extensive error
 message indicating the available and required scopes.
 
+Arrays of required scopes can also be templated in the scopes section. When
+this form is used, the parameter must be an array.
+
+```js
+  scopes: {AllOf: [
+    {for: 'route', in: 'routes', each: 'queue:route:<route>'}
+  ]}
+```
+
+Given a parameter `routes` that is an array of strings as in `['foo', 'bar']`,
+this will evaluate to:
+
+```js
+  scopes: {AllOf: ['queue:route:foo', 'queue:route:bar']}
+```
+
+You may also use if/then constructs in scope expressions. In this case, the `if`
+field should be a parameter and the `then` must be a scope expression that will
+be subsituted in for the if/then block if the `if` parameter exists and is truthy.
+Given the follwoing example:
+
+```js
+  scopes: {AnyOf: ['foo:bar', {if: 'public', then: {AnyOf: []}}]}
+```
+
+We can call `authorize({public: true})` and the method call will be permitted
+even if the client does not have the scope `foo:bar`.
+
+In addition, parameters can be nested objects and accessed with dotted notation.
+Given parameters `task` that is `{extra: {treeherder: {symbol: 'F'}}}`, we can do
```

Comment:
Is this smart? 

My idea was that the parameters should be: `{'task.extra.treeherder.symbol': 'F'}` instead....

Note: I'm 100% unsure of all that we're doing here...

Reply:
I think what you're saying here is, let's don't use `_.at` and instead require all parameters to be named -- but allowing dots in parameters if the user prefers.  I can support that.  `req.params` will never have structure -- it's just strings.  And anything else is provided by the API implementation.

Reply:
My thinking with `_.at` is to allow us to just pass the task in directly to `authorize` and grab `task.scopes` or `task.routes` out of it directly. This seems like a common use-case?

Reply:
It doesn't seem to make things too much more complex.  Probably good to point to the `_.at` docs to indicate that it can also index array.